src/metabase/api/dashboard.clj

```diff
@@ -792,6 +866,22 @@
                                 :present #{:description :position :width :collection_id :collection_position :cache_ttl :archived_directly}
                                 :non-nil #{:name :parameters :caveats :points_of_interest :show_in_getting_started :enable_embedding
                                            :embedding_params :archived :auto_apply_filters}))]
+             (when (api/column-will-change? :archived current-dash dash-updates)
+               (if (:archived dash-updates)
+                 (card/with-allowed-changes-to-internal-dashboard-card
+                   (t2/update! :model/Card
+                               :dashboard_id id
+                               :archived false
+                               {:archived true :archived_directly false}))
+                 (card/with-allowed-changes-to-internal-dashboard-card
+                   (t2/update! :model/Card
+                               :dashboard_id id
+                               :archived true
+                               :archived_directly false
+                               {:archived false}))))
```

Comment:
since you wrap both branches of this if expression w/ `card/with-allowed-changes-to-internal-dashboard-card`, you could wrap the if with it instead. Might be clearer to wrap the exact update statements though, so I don't have a strong conviction here.